### PR TITLE
HIVE-26202: Refactor Iceberg Writers

### PR DESCRIPTION
**Pull request by pvary**
### What changes were proposed in this pull request?
We started to have too many writers and the creation process was hard to understand.
Created a WriterBuilder with a clean interface to hide the internals

### Why are the changes needed?
Make it easier to read the code

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Unit tests

### CODE REVIEW

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputFormat.java [hunk 1]
```diff
@@ -63,39 +60,19 @@ public void checkOutputSpecs(FileSystem ignored, JobConf job) {
...
-      HiveFileWriterFactory writerFactory = new HiveFileWriterFactory(table, fileFormat, schema, null, fileFormat,
-          null, null, null, schema);
-      return new HiveIcebergRecordWriter(schema, table.specs(), table.spec().specId(), fileFormat, writerFactory,
-          outputFileFactory, io, targetFileSize, taskAttemptID, tableName, false);
-    }
+    int poolSize = jc.getInt(DELETE_FILE_THREAD_POOL_SIZE, DELETE_FILE_THREAD_POOL_SIZE_DEFAULT);
+
+    return WriterBuilder.builderFor(table)
+        .queryId(jc.get(HiveConf.ConfVars.HIVEQUERYID.varname))
+        .tableName(tableName)
```
[the remark this answers is not in the document]
Comment (pvary) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputFormat.java:72]: AFAIK there is no tableName for the table object.
Do I miss something?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputFormat.java [hunk 2]
```diff
@@ -32,19 +31,17 @@
...
-import org.apache.iceberg.io.OutputFileFactory;
 import org.apache.iceberg.mr.Catalogs;
+import org.apache.iceberg.mr.hive.writer.HiveIcebergWriter;
+import org.apache.iceberg.mr.hive.writer.WriterBuilder;
 import org.apache.iceberg.mr.mapred.Container;
-import org.apache.iceberg.util.PropertyUtil;
 
 public class HiveIcebergOutputFormat<T> implements OutputFormat<NullWritable, Container<Record>>,
     HiveOutputFormat<NullWritable, Container<Record>> {
+  private static final String DELETE_FILE_THREAD_POOL_SIZE = "iceberg.delete.file.thread.pool.size";
```
[the remark this answers is not in the document]
Comment (pvary) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputFormat.java:43]: We should definitely document it in our upcoming documentation

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputFormat.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputFormat.java:72]: Sorry, my mistake. I confused with table.name() which returns `catalogName.dbName.tblName`

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 3]
```diff
@@ -571,6 +571,15 @@ public static boolean isUpdate(Configuration conf, String tableName) {
         conf.get(InputFormatConfig.OPERATION_TYPE_PREFIX + tableName));
   }
 
+  public static Operation operation(Configuration conf, String tableName) {
```
[the remark this answers is not in the document]
Comment (pvary) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:559]: I did this. Could you please check the result?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/writer/WriterBuilder.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.OutputFileFactory;
+import org.apache.iceberg.util.PropertyUtil;
+
+import static org.apache.iceberg.TableProperties.DEFAULT_FILE_FORMAT;
+import static org.apache.iceberg.TableProperties.DEFAULT_FILE_FORMAT_DEFAULT;
+import static org.apache.iceberg.TableProperties.DELETE_DEFAULT_FILE_FORMAT;
+
+public class WriterBuilder {
+  private Table table;
```
[the remark this answers is not in the document]
Comment (pvary) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/writer/WriterBuilder.java]: Done

**Review by lcspinter — APPROVED** [review]
+1

**Pull request merged**